File: FBAudio-Android/app/src/main/java/com/fba/app/domain/model/Content.kt
package com.fba.app.domain.model

import android.net.Uri

/** One entry of the website's curated side menu (collections, themes, people, places…). */
data class MenuNode(
    val label: String,
    val link: String,
    val om: Boolean = false,
    val children: List<MenuNode> = emptyList(),
) {
    val hasChildren: Boolean get() = children.isNotEmpty()
    val isExternal: Boolean get() = link.startsWith("http") && !link.contains("freebuddhistaudio.com")
    val isPlaceholder: Boolean get() = link.isBlank() || link.startsWith("#")

    /** The /collection/<slug> slug when this entry is a named collection. */
    val collectionSlug: String?
        get() = Regex("/collection/([^/?#]+)").find(link)?.groupValues?.get(1)?.trim()

    /** A [ContentSource] for this entry, or null when it is a placeholder/external link. */
    fun toSource(): ContentSource? {
        val slug = collectionSlug
        val path = link.trim().removePrefix("https://www.freebuddhistaudio.com")
        return when {
            slug != null -> ContentSource.NamedCollection(slug)
            path.startsWith("/series/details") -> ContentSource.Series(path)
            path.startsWith("/browse") || path.startsWith("browse") -> {
                // "/browse?cat=speakers&t=audio" style index pages are the API collections
                val cat = Regex("[?&]cat=([^&]+)").find(path)?.groupValues?.get(1)
                if (cat != null) ContentSource.ApiCollection(cat, label)
                else ContentSource.Browse("/" + path.removePrefix("/"))
            }
            path.startsWith("/audio/details") -> null
            else -> null
        }
    }
}

/**
 * Where a list of talks/series comes from. Encoded into navigation routes with
 * [encode] and decoded with [decode]; keep it a flat string-friendly shape.
 */
sealed class ContentSource {
    /** `/api/v1/collections/{type}` — latest, introductions, speakers, all_series… */
    data class ApiCollection(val type: String, val title: String = "") : ContentSource()
    /** A curated `/collection/{slug}` page (themes, collections, some people/places). */
    data class NamedCollection(val slug: String) : ContentSource()
    /** A `/browse?…` listing: one speaker, place, year or genre. */
    data class Browse(val path: String) : ContentSource()
    /** A `/series/details?num=…` page. */
    data class Series(val path: String) : ContentSource()

    fun encode(): String = when (this) {
        is ApiCollection -> "api|$type|$title"
        is NamedCollection -> "named|$slug|"
        is Browse -> "browse|$path|"
        is Series -> "series|$path|"
    }

    companion object {
        fun decode(encoded: String): ContentSource? {
            val parts = encoded.split('|', limit = 3)
            if (parts.size < 2) return null
            val arg = parts[1]
            val extra = parts.getOrElse(2) { "" }
            return when (parts[0]) {
                "api" -> ApiCollection(arg, extra)
                "named" -> NamedCollection(arg)
                "browse" -> Browse(arg)
                "series" -> Series(arg)
                else -> null
            }
        }

        fun seriesByCatNum(catNum: String) = Series("/series/details?num=${Uri.encode(catNum)}")
    }
}

/** One page of a list plus what's needed to show its header and fetch the next page. */
data class ListPage(
    val items: List<SearchResult>,
    val totalItems: Int,
    val page: Int,
    val title: String = "",
    val description: String = "",
    val imageUrl: String = "",
    val hasRemaster: Boolean = false,
    val omOnly: Boolean = false,
    /** For browse listings: the API endpoint + query that pages the same list. */
    val apiUrl: String = "",
    val apiQuery: String = "",
) {
    val hasMore: Boolean get() = items.isNotEmpty() && totalItems > 0 && page * PAGE_SIZE < totalItems

    companion object {
        const val PAGE_SIZE = 24
    }
}

/** Content of the website's Digital Legacy page. */
data class DigitalLegacy(
    val title: String,
    val description: String,
    val sampleCatNum: String,
    val seriesPath: String = "/series/details?num=X16",
)

File: FBAudio-iOS/FBAudio/Models/Content.swift
import Foundation

/// One entry of the website's curated side menu (collections, themes, people, places…).
struct MenuNode: Codable, Equatable, Hashable, Identifiable {
    let label: String
    let link: String
    let om: Bool
    let children: [MenuNode]

    var id: String { "\(label)|\(link)" }

    init(label: String, link: String, om: Bool = false, children: [MenuNode] = []) {
        self.label = label
        self.link = link
        self.om = om
        self.children = children
    }

    var hasChildren: Bool { !children.isEmpty }
    var isExternal: Bool { link.hasPrefix("http") && !link.contains("freebuddhistaudio.com") }
    var isPlaceholder: Bool { link.isEmpty || link.hasPrefix("#") }

    /// The /collection/<slug> slug when this entry is a named collection.
    var collectionSlug: String? {
        guard let range = link.range(of: "/collection/") else { return nil }
        let rest = link[range.upperBound...]
        let slug = rest.prefix { $0 != "/" && $0 != "?" && $0 != "#" }
        let trimmed = slug.trimmingCharacters(in: .whitespaces)
        return trimmed.isEmpty ? nil : trimmed
    }

    /// A ContentSource for this entry, or nil when it is a placeholder/external link.
    func toSource() -> ContentSource? {
        if let slug = collectionSlug { return .namedCollection(slug) }
        var path = link.trimmingCharacters(in: .whitespaces)
        if path.hasPrefix("https://www.freebuddhistaudio.com") {
            path = String(path.dropFirst("https://www.freebuddhistaudio.com".count))
        }
        if path.hasPrefix("/series/details") { return .series(path) }
        if path.hasPrefix("/browse") || path.hasPrefix("browse") {
            if let cat = ContentSource.queryValue("cat", in: path) {
                return .apiCollection(cat, title: label)
            }
            return .browse(path.hasPrefix("/") ? path : "/" + path)
        }
        return nil
    }
}

/// Where a list of talks/series comes from. String-encodable for navigation.
enum ContentSource: Hashable, Codable {
    /// `/api/v1/collections/{type}` — latest, introductions, speakers, all_series…
    case apiCollection(String, title: String = "")
    /// A curated `/collection/{slug}` page (themes, collections, some people/places).
    case namedCollection(String)
    /// A `/browse?…` listing: one speaker, place, year or genre.
    case browse(String)
    /// A `/series/details?num=…` page.
    case series(String)

    var isSeries: Bool { if case .series = self { return true } else { return false } }

    func encode() -> String {
        switch self {
        case .apiCollection(let type, let title): return "api|\(type)|\(title)"
        case .namedCollection(let slug): return "named|\(slug)|"
        case .browse(let path): return "browse|\(path)|"
        case .series(let path): return "series|\(path)|"
        }
    }

    static func decode(_ encoded: String) -> ContentSource? {
        let parts = encoded.split(separator: "|", maxSplits: 2, omittingEmptySubsequences: false).map(String.init)
        guard parts.count >= 2 else { return nil }
        let arg = parts[1]
        let extra = parts.count > 2 ? parts[2] : ""
        switch parts[0] {
        case "api": return .apiCollection(arg, title: extra)
        case "named": return .namedCollection(arg)
        case "browse": return .browse(arg)
        case "series": return .series(arg)
        default: return nil
        }
    }

    static func seriesByCatNum(_ catNum: String) -> ContentSource {
        .series("/series/details?num=\(catNum.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? catNum)")
    }

    /// Value of a query parameter in a path like "/browse?cat=latest&t=audio".
    static func queryValue(_ key: String, in path: String) -> String? {
        guard let q = path.split(separator: "?", maxSplits: 1).dropFirst().first else { return nil }
        for pair in q.split(separator: "&") {
            let kv = pair.split(separator: "=", maxSplits: 1).map(String.init)
            if kv.count == 2, kv[0] == key { return kv[1] }
        }
        return nil
    }
}

/// One page of a list plus what's needed to show its header and fetch the next page.
struct ListPage: Codable {
    static let pageSize = 24

    var items: [SearchResult]
    var totalItems: Int
    var page: Int
    var title: String = ""
    var description: String = ""
    var imageUrl: String = ""
    var hasRemaster: Bool = false
    var omOnly: Bool = false
    /// For browse listings: the API endpoint + query that pages the same list.
    var apiUrl: String = ""
    var apiQuery: String = ""

    var hasMore: Bool { !items.isEmpty && totalItems > 0 && page * ListPage.pageSize < totalItems }
}

/// Content of the website's Digital Legacy page.
struct DigitalLegacy: Codable {
    let title: String
    let description: String
    let sampleCatNum: String
    var seriesPath: String = "/series/details?num=X16"
}
